[Lee Wai Chong] iP

leewaichong opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen [source](https://dansilvestre.com/productivity-quotes/)

DukePro frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is FREE!

Features:

- [x] Managing tasks

- [ ] Managing deadlines (coming soon)

- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Commit:
Level-2

Commit:
Level-3

Commit:
Level-4

Commit:
Level-6

Commit:
Level-5

Commit:
Level-1

Commit:
A-TextUiTesting

Commit:
save

Commit:
Level-7

Commit:
Level-8

Commit:
Merge branch 'Branch-Level-7'

Commit:
Merge branch 'Branch-Level-8'

Commit:
Fix merge bugs

Commit:
Fix-Level-8-Bugs

Commit:
A-MoreOOP

Commit:
A-Packages

Commit:
Fixed-Minor-inconsistencies

Commit:
Merge remote-tracking branch 'upstream/add-gradle-support'

cowlinn commented on src/main/java/duke/Parser.java:
```diff
@@ -18,15 +20,15 @@ else if (str.split(" ")[0].equals("unmark")) {
                     storage.writeToFile(taskList);
                 }
-                else if(str.split(" ", 2)[0].equals("todo")) {
+                else if(str.split(" ", 2)[0].equals("duke.todo")) {
```
Some minor improvement suggestions (feel free to use as many or as little as you'd like xd)
1.  Perhaps give `str.split(" " ,2)[0] `  a better name? Something like `header`
2.  Maybe can try using a switch statement if you're just doing comparisons?

cowlinn commented on src/main/java/duke/Parser.java:
```diff
@@ -1,3 +1,5 @@
+
 import java.io.IOException;
 public class Parser {
     public static boolean parse(String str, TaskList taskList, Ui ui, Storage storage) throws DukeException {
```
Perhaps to make the method name shorter, you could consider wrapping the whole `parse` method in a try-catch? So you don't have to explicitly mention that it throws `DukeException ` sooo something like 

```
public static boolean parse(String str, TaskList taskList, Ui ui, Storage storage) { 
    try { //main logic 
     
    } catch (DukeException d) {
    // handle the exception here 
    } catch (moreExceptionsHere m) { 
    // gives you the freedom to handle more exceptions down the line 
```

cowlinn commented on src/main/java/duke/Task.java:
```diff
@@ -19,31 +21,31 @@ public void unmark() {
 
     /**
-     * use todo, event, deadlines
+     * use duke.todo, duke.event, deadlines
```
Ay good use of enums! I like that it isn't just a string :)

cowlinn commented on src/main/java/duke/Storage.java:
```diff
@@ -1,3 +1,5 @@
+package duke;
+
 import java.io.*;
```
Perhaps could try to minimize wildcard imports? To fix this fast, delete this statement, and right click all the red squiggly lines and let intelli-j auto import for you.
IntelliJ might automatically wildcard import if you have more than 3 classes, so need to uncheck that option

cowlinn commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,75 @@
+package duke;
+
+import java.io.*;
```
Perhaps could try to minimize wildcard imports? To fix this fast, delete this statement, and right click all the red squiggly lines and let intelli-j auto import for you.
IntelliJ might automatically wildcard import if you have more than 3 classes, so need to uncheck that option

cowlinn commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,45 @@
+import java.io.IOException;
+public class Parser {
+    public static boolean parse(String str, TaskList taskList, Ui ui, Storage storage) throws DukeException {
```
Perhaps to make the method name shorter, you could consider wrapping the whole parse method in a try-catch? So you don't have to explicitly mention that it throws DukeException sooo something like

```
public static boolean parse(String str, TaskList taskList, Ui ui, Storage storage) { 
    try { //main logic 
     
    } catch (DukeException d) {
    // handle the exception here 
    } catch (moreExceptionsHere m) { 
    // gives you the freedom to handle more exceptions down the line 
```
Edit* : Sorry I just realized you were alr doing this i think can remove the `throws dukeException` from the code and it'll be okay alr

cowlinn commented on src/main/java/duke.txt:
```diff
@@ -1 +1,2 @@
-[E][ ] your father (at: Sep 9 2018)
+[T][ ] your mother
```
Maybe can add the save file into gitIgnore! If not it'll keep popping up in your commits unless you wanna save your changes

cowlinn commented on src/main/java/duke/Parser.java:
```diff
@@ -20,15 +20,15 @@ else if (str.split(" ")[0].equals("unmark")) {
                     storage.writeToFile(taskList);
                 }
-                else if(str.split(" ", 2)[0].equals("duke.todo")) {
+                else if(str.split(" ", 2)[0].equals("todo")) {
```
Small improvements you could consider! Free free to ignore they are mostly style related
1. Give a name to` str.split(" ", 2)[0]`, maybe call it header or something more meaningful
2.  Can consider using switch statement since you're checking for equality throughout

cowlinn commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,57 @@
+     * @param tasks
+     * @param str
+     */
+    public static void makeTask(TaskList tasks, String str) throws DukeException {
```
Ay good use of enums! I like that it isn't just a string :)

cowlinn commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.Scanner;
+public class Ui {
+    private Scanner sc;
+    private static final String INTRO = "Hi! I'm ELON MUSK\nWhat can I do for you?";
```
Great that you've abstracted print lines into the Ui class! To prevent long sentences, perhaps you can shorten it via string concatenation into a new line, so something like their duke logo. 
![Capture](https://user-images.githubusercontent.com/88648894/187116468-6fb6a756-3039-4922-a70e-098d7909aa09.JPG)

I'm not sure that using System.out.println in multiple classes is the best idea? Maybe consider using a Stringbuilder instead of printing to console, so you only do one final `system.out.println`. Might help for this week's task, where you have to consolidate ElonMusk's response into one string

cowlinn commented on src/main/java/Storage.java:
```diff
@@ -19,18 +18,19 @@ public static void writeToFile(ArrayList<Task> tasks) throws IOException {
 
         while((str = reader.readLine()) != null) {
+
             if(str.charAt(1) == 'T') {
```
Perhaps try to be consistent when using enums? Maybe save 'T' / 'X', 'D', etc. as some sort of character enum

cowlinn commented on src/main/java/Storage.java:
```diff
@@ -45,15 +45,8 @@ else if(str.charAt(1) == 'D') {
+                String result = input.correctDateFormat(str.substring(openBracket + 5, closeBracket));
+                input.setDate(new formatDate(result));
 
                 if (str.charAt(4) == 'X') {
```
Maybe I could ask why these checks are repeated / done in storage repeatedly? The way I might do it would be to create a static factory class within each of deadline/todo/event, and call an appropriate constructor depending on whether it is 'E', 'T' or 'D'. 

In general, I think you could consider trying to abstract out the parsing of the saveData and creation into a seperate parser class / within each Event/Deadline/Task class itself, and not attempt to create it within Storage :D Could save on repetition of code, and it might make use of better OOP principles!

cowlinn reviewed (commented):
Heyo great job just some minor inconsistencies here and there. You might also wanna make sure your classes are named Properly (e.g: Event/Deadline/Task should all be capitalized), and use `checkstyle` to ensure any small spacing errors are resolved. Lmk if you got any questions regarding my comments!

Commit:
A-JUnit

Commit:
A-Jar

Commit:
A-JavaDoc

Commit:
Fix merge bugs

Commit:
Fix merge bugs

phyliciachristel commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    /**
+     *
```
Perhaps you should include a description with the summary of your method before the parameter section? I noticed this in other parts of your code as well

phyliciachristel commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,59 @@
+     */
+    public static boolean isParse(String str, TaskList taskList, Ui ui, Storage storage) throws DukeException {
+
+            try {
```
why is your code indented by 8 spaces over here?

phyliciachristel commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,59 @@
+                else if (str.equals("list")) {
+                    ui.printList(taskList);
+                }
+                else if (str.split(" ")[0].equals("mark")) {
```
would it be better if you called `str.split(" ")[0]` and `str.split(" ", 2)[0]` at the start of your try block to avoid code duplication and repeated computation of the same line?

phyliciachristel commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,61 @@
+     * @param tasks add task into this taskList
+     * @param input prints out the task added
+     */
+    public static void addedTask(TaskList tasks, Task input) {
```
would addTask be a better name since you are currently adding the task in this method?

phyliciachristel commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,94 @@
+
+        while((str = reader.readLine()) != null) {
+
+            if(str.charAt(1) == 'T') {
```
Would it be better if you computed `str.charAt(1)` before the if-else block? Would it be simpler if you implemented a switch case instead?

phyliciachristel commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,52 @@
+
+    /**
+     *
+     *@param str "MMM dd yyyy"
```
Perhaps you should add a space after * to comply with the coding guidelines

mjgui commented on src/main/java/duke.txt:
```diff
@@ -0,0 +1,7 @@
+[E][X] project (at: Sep 9 2019)
```
As mentioned by cowlinn, perhaps consider including your output file in the `.gitignore`? This will prevent unnecessary changes to your commits

mjgui commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,52 @@
+public class Deadline extends Task {
+    protected FormatDate date;
+
+    /**
+     *
+     * @param description description of task
+     */
```
On top of what phyliciachristel mentioned, maybe capitalise your descriptions and end them off with a full stop, as per the guidelines [here](url)?

mjgui commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,11 @@
+public class DukeException extends Exception {
+    /**
+     *
+     * @param message shows error message for DukeException
```
Maybe separate the description of the method (storing the error message) from the purpose of the parameter (error message itself)? Since the exception does not do any showing of the error message.

mjgui commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,59 @@
+                if (str.equals("bye")) {
+                    ui.bye();
+                    return true;
+                }
+                else if (str.equals("list")) {
```
Aside from comments by other reviewers, perhaps you should consider putting your closing curly brace for the `if` statement on the same line as `else {` to conform to the [coding guidelines](https://se-education.org/guides/conventions/java/intermediate.html#layout)?

mjgui commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,94 @@
+     * @throws IOException
+     */
+    public void writeToFile(TaskList tasks) throws IOException {
+       //file = new File("src/main/java/duke.txt");
```
Should this comment be removed since it is unused?

mjgui commented on text-ui-test/EXPECTED.TXT:
```diff
@@ -1,7 +1,7 @@
-| | | | | | | |/ / _ \
-| |_| | |_| |   <  __/
-|____/ \__,_|_|\_\___|
+Hello! I'm duke.Duke.
```
This was probably introduced by IntelliJ's search and replace - perhaps consider changing this to your actual expected text?

mjgui reviewed (commented):
Good job overall, left some comments regarding coding standards

Commit:
Fix merge bugs

Commit:
Revert to Commit A-Jar

Commit:
merge Branch-A-JavaDoc

Commit:
merge Branch-Level-9

Commit:
Merge branch 'Branch-A-CodingStandard'

Commit:
fixing merge bugs

Commit:
Level-10

Commit:
Merge tag 'Branch-Level-10'

Commit:
Add Assertions

Logics are added to ensure main methods work well.
Methods like delete, unmark and mark have assert methods to ensure the list indexing is correct and that the logic holds.

Commit:
Improve CodeQuality

Made Code in File Parser shorter by using SLAP.
Improved Readability of the code.

leewaichong committed:
Merge pull request #3 from Whysochong/Branch-A-CodeQuality

Improve CodeQuality

leewaichong committed:
Merge pull request #2 from Whysochong/Branch-A-Assertions

Add Assertions

Commit:
Minor Changes in Code

A missing piece of Assert Statement was added.

Commit:
BCD-Extensions & Minor Code Glitches Fixed

Did and completed the new feature Snooze/Reschedule.
Fixed the Glitch where actions/features are not written to file after executing.

Commit:
Minor changes (Bugs)

Changed small glitch and bugs with Todo.toString().

Commit:
Minor changes (Bugs)

Changed small glitch and bugs with Todo.toString().

Commit:
Minor changes (Bugs)

Changed small glitch and bugs with Todo.toString().

Commit:
Upload picture and Complete User Guide

Commit:
A-UserGuide

Commit:
fix-mkdir-bugs
